enhancements/cluster-logging/elasticsearch-cert-management.md

```diff
@@ -0,0 +1,158 @@
+
+Currently our managed Elasticsearch cluster requires anything using our operator to generate, manage, and provide certs and secrets for the operator and its cluster to use. This leads to many potential sources to create certificates and can cause a gap between what the managed Elasticsearch requires and what something may provide. This would instead make the Elasticsearch Operator a single source of truth for how certificates would be generated.
+
+### Goals
+The specific goals of this proposal are:
+
+* Outline the responsibility for the Elasticsearch Operator regarding certificates
+
+* Expose the desired annotation that the operator would watch for on secrets where an elasticsearch CR exists
+
+* Discuss upgrade strategies from the current LTS mechanism.
+
+We will be successful when:
+
+* Users of Elasticsearch Operator no longer need to maintain certificates for Elasticsearch and can have certificates injected into their secrets
+
+* There is a clear path forward for automatic upgrades
+
+### Non-Goals
+
+* This does not seek to implement work to adhere to the cluster level TLS definitions, however this does make the implementation less complex for the Elasticsearch use-case.
+
+## Proposal
+
+Instead of the current implementation where any consumer of the Elasticsearch Operator is responsible for providing certificates and secrets for Elasticsearch and the operator to use, we want to shift this responsibility to the Elasticsearch Operator instead. This will greatly improve the use case of the operator and make the operator the single source of truth for its certificates.
+
+To allow other components to still be able to communicate with the managed cluster via mTLS the operator will watch for a specific annotation on secrets in the namespace where an elasticsearch CR is present and will inject into this secret.
+
+### User Stories
+
+#### As an user of EO, I do not want to have to generate and provide secrets for the Elasticsearch cluster
+
+We want to further simplify using the Elasticsearch Operator and create a single source of truth for the certs the cluster uses.
+
+#### As a consumer of EO, I want a mechanism to have certificates injected into a secret for me to continue mTLS communcation with my Elasticsearch cluster
+
+We want to continue allowing mTLS as this supports the Cluster Logging and Jaeger use cases, however they shouldn't need to do more than notify they need certs injected by EO.
+
+### Implementation Details
+
+#### Assumptions
+
+* The Elasticsearch Operator will implement this in golang (rather than the current script used in CLO)
+
+* The operator will need to be able to store generated certificates for each namespace/unique cluster so that there is no security concerns of clusters cross communicating.
+
+#### Security
+
+Ensuring that signing certificates and configurations are isolated for different clusters is a must for this implementation. It will also need to be able to do this thread safe manner.
```

Comment:
"in a thread safe manner". Thread safety is not really a security concern though.

Reply:
can't it be if we're reading in certificates that we've generated for another cluster? we want to ensure things are completely isolated

Reply:
I thought under thread safety you meant absence of race conditions or deadlocks. Can you help me understand how k8s objects from a namespace and a thread in the operator are related? On the surface it seems it should be enough during reconcile to look at the namespace the objects being reconciled are from and keep things isolated based on that. On the other hand, would running an EO instance per Elastic cluster alleviate such concerns? I'm sure I'm missing important details.

Reply:
my thought was given the Elasticsearch operator can process es events from multiple namespaces it may be possible that it is trying to resolve two different namespaces at the same time. so in that case it would be processing certificates "at the same time". this may not end up being a valid concern, but it was my original thought behind the original statement.